### PR TITLE
[Merged by Bors] - chore: Add a test implementation of bors

### PR DESCRIPTION
**Pull request by jamtur01**
This PR enables https://bors.tech for our PR merge workflow. Now we run an abbreviated lint-centric check on the PR rather than the full test suite. We've achieved this by running the lint tests inside a container running the `timberio/ci_image` Docker image. 

We also consolidated and updated the image build process in `environment.yml` to keep the `ci_image` up-to-date.

Then, instead of pressing merge, we use the `bors` commands to submit the PR to a `staging` branch where CI is run. The new CI run merges the prior e2e and tests jobs as a group. If these return green then the `staging` branch is fast-forwarded/merged into `master`.

Signed-off-by: James Turnbull <james@lovedthanlost.net>

### CODE REVIEW

**Comment by jamtur01** [comment]
bors try

**Comment by jamtur01** [comment]
bors try

**Comment by jamtur01** [comment]
bors try

**Comment by MOZGIII** [comment]
One simpler alternative I was thinking about is the Github feature of disallowing the merge unless checks/tests are passed. This doesn't allow offloading the tests though. But it has a benefit of not requiring to manually kick-off the tests when you need just that (and not the whole merge).

**Comment by MOZGIII** [comment]
Another alternative - instead of queueing up everything on the `staging` branch, we can have a command that would invoke the jobs programmatically via the `dispatch_workflow` API call (if needed - multiple stages in the chain).

**Comment by jamtur01** [comment]
I don't think either of those solves our merge-fast bottleneck problem?

**Comment by jamtur01** [comment]
bors try

**Comment by MOZGIII** [comment]
Could you please outline the motivation for this change? The merge-fast bottleneck problem - what that is exactly?

I agree that bors does a lot for us, but it might be that we'd want to use bors *and* take pieces from both things I mentioned above - the first one will allow disabling the merge button, and the second one should allow splitting the workflows (not sure if it's necessary - maybe the motivation will make it clearer).

The expected outcome of running bors is, I presume, the reduced about of workflow runs - we'll only be running them on merge or manual request - and that would give us an increase in volume - correct? And it will also provide us more guarantees - in a sense that test will be guaranteed to pass before the automatic merge passes?

**Comment by jamtur01** [comment]
@MOZGIII This is the original ticket - https://github.com/timberio/vector/issues/2074 - re-opened by @JeanMertz who probably has thoughts on this.

**Comment by binarylogic** [comment]
@MOZGIII to my understanding this is about productivity. PRs go through a fast set of unit tests, and once green, we merge into a "merge" branch that passively runs longer comprehensive tests. This way we can reduce the time spent waiting on PRs while still having some confidence in the code. @JeanMertz does that sound correct?

**Comment by MOZGIII** [comment]
Thx, I'll check it out.

From my prior experience, any transition to the system that conditions merges on the passing tests requires solving the test flakiness problem completely and reliably. If we want to switch to bors, or use the Github feature to block merge until the tests pass - we should consider being hard-blocked on flaky tests. If we transition and the tests won't be passing reliably - the work will probably stall significantly, and a recovery process will be required before we can continue.
To sum up - my point here is that we have to be careful and not rush this.

**Comment by MOZGIII** [comment]
> @MOZGIII to my understanding this is about productivity. PRs go through a fast set of unit tests, and once green, we merge into a "merge" branch that passively runs longer comprehensive tests before going into `master`. This way we can reduce the time spent waiting on PRs while still having some confidence in the code. @JeanMertz does that sound correct?

I don't see how we wait on PRs less. The overall time spent might increase due to queuing on branches. The only benefit I see there is automerge - it's very easy to implement in our current setup though.

**Comment by jamtur01** [comment]
bors try

**Comment by MOZGIII** [comment]
So, the optimization that a system like bors can provide (and might already be doing this) is to decrease the overall amount of tests to be lower than PRs, by combining the changes multiple PRs together and running the tests for this buck once. Then, if everything goes smooth - changes are promoted, if something goes wrong - the bad PR is searched for by running the tests on each PR individually - and the failure is then reported at the relevant PR.
The problem here is for a person waiting on tests to pass, for the failing PRs the wait time doubles - as there are at least two test runs to report that the PR is failing the tests.


One thing I like about the concept is that we should be able to remove the conditioning on what k8s tests to run, and just run all of them - because due to restricting the tests to just two branches we'll be only using very few Github Action runner slots.

**Comment by MOZGIII** [comment]
We should probably move the discussion to the issue, as the comments of this PR will likely to be used for testing bors :smile:

**Comment by JeanMertz** [comment]
> One simpler alternative I was thinking about is the Github feature of disallowing the merge unless checks/tests are passed.

I agree with @jamtur01 that this doesn't solve any of the issues we're trying to solve, and in fact, I would even say it will make things worse. We're all adults, so I don't think forcing us to wait for long CI runs if you have a good reason not to, is a good practice to follow. If we had a faster CI workflow (which bors _might_ bring us), then I think it's actually not such a terrible idea.

> we should consider being hard-blocked on flaky tests.

I agree, and I think we should just disable any flaky test, even before we look into fixing the test. Disabling a test is a 5-minute job, and prevents many issues down the line. Obviously, the test itself is there for a reason, so we should still plan to improve the test and re-enable it, but it shouldn't block any other ongoing work.

> I don't see how we wait on PRs less. The overall time spent might increase due to queuing on branches.

The time it takes to get the changes into `master` won't change, but waiting on PRs definitely will, since Bors will handle the longer test runs, and will merge multiple PRs simultaneously if need be.

> @MOZGIII to my understanding this is about productivity.

What @binarylogic wrote is correct. I would say there are three upsides to a system like Bors:

1. We can move longer running behaviour/integration tests into the merge PR, which is managed by Bors and thus doesn't require our attention. This will speed up CI runs for the PRs we ourselves create.
2. Once reviewers are happy with the PR, you can ask Bors to auto-merge once the CI is green, this means you are now "done" with the PR, and don't have to check back every so often to see if CI passed before you can click the "merge" button.
3. For external contributors, we can give them "merge rights" for each individual PR once the reviewer is happy with the changes, except for a few minor nits. Meaning, if you want them to change a variable name, or update some other parts of the PR that doesn't significantly impact the functionality of the code, you can ask them to do so, and tell Bors to accept any merge request from the PR author. This again allows you to move on to other work, without having to circle back to approve the final merge.

In general, I've seen very positive results from Bors-managed repositories. However, the major blocker might still be the lack of a useable merge-commit workflow that fits our current model, something @Hoverbear mentioned might still not be possible with Bors.

File: CONTRIBUTING.md
Comment (JeanMertz) [inline CONTRIBUTING.md:213]: There's also `bors delegate+` (or `bors d+`), which allows reviewers to delegate `bors r+` to the PR author.

This allows us to review, ask for some minor changes, and tell the author that they can ask Bors to merge themselves, once they've addressed the feedback.

You wouldn't do this for PRs that require significant changes or more discussions, but for styling-only change requests, it will free you of having to circle back to the PR to `bors r+`, and gives more authority to PR authors when we think they can use that authority wisely.

**Review by JeanMertz — COMMENTED** [review]
I wonder if we still want to run some unit tests within the PR itself?

Perhaps we wouldn't run them for all platforms, but at least the x86 unit tests might be fast enough to run, which gives you a reasonably high confidence that things are working as expected, before Bors runs the full suite on the staging branch?

**Comment by jamtur01** [comment]
@JeanMertz Yes - I started with the bare minimum and was going to work out what was an acceptable base. Will add in the Linux unit tests too.

File: CONTRIBUTING.md
Comment (jamtur01) [reply CONTRIBUTING.md:213]: Yeah - I skipped that initially in the docs as a larger discussion on contributor workflow.

**Comment by MOZGIII** [comment]
Well, I'm curious to see how bors will work for us, don't get me wrong here.

> 1. We can move longer running behaviour/integration tests into the merge PR, which is managed by Bors and thus doesn't require our attention. This will speed up CI runs for the PRs we ourselves create.

How does it not require our attention? If the CI is green, this is no different from the automerge - i.e. as in GitLab, you can tick the checkbox to wait for green CI and press the merge button - and it'll merge if the CI passes after that, or abort if the CI fails.

What we have to consider here is the failure scenario. Is it still acceptable to rely on CI to run the full test suite, rather than running the tests locally? If so, the time-to-failure-report matters, and I wonder if my concerns at https://github.com/timberio/vector/pull/4178#issuecomment-700416506 are what we'll see in practice.

I've worked on PRs to rust project before, so I know bors can be very slow, however I wasn't able to compare the speed without it. :smile: 

> 2. Once reviewers are happy with the PR, you can ask Bors to auto-merge once the CI is green, this means you are now "done" with the PR, and don't have to check back every so often to see if CI passed before you can click the "merge" button.

Auto-merge (by itself) is great! Definitely, something I'd like to have. Not for or against having it as part of bors integration.

> 3. For external contributors, we can give them "merge rights" for each individual PR once the reviewer is happy with the changes, except for a few minor nits. Meaning, if you want them to change a variable name, or update some other parts of the PR that doesn't significantly impact the functionality of the code, you can ask them to do so, and tell Bors to accept any merge request from the PR author. This again allows you to move on to other work, without having to circle back to approve the final merge.

This is very nice indeed, something I didn't think about. It's a nice-to-have though, and might not worth it if the integration speed degrades.

**Comment by Hoverbear** [comment]
I'm very happy to use bors or not and don't have a strong opinion here. It's a good tool if we want to use it the way it's meant to be used.

**Comment by JeanMertz** [comment]
I'll leave @jamtur01 to run their experiment, which I'm excited to see happen 👍

I don't think debating hypotheticals is particularly useful at this stage. Let's see how this PR turns out, then I would vote to merge and trial this workflow, and if we don't like it, we can always back out of it, no harm done.

Whatever we end up with, the current situation is something we want to fix, I think we can all agree on that. This is _one_ way to try and fix that situation, it _seems_ like it would help us, so a trial period makes sense to me, if it doesn't, we'll move on to a different potential solution.

**Comment by MOZGIII** [comment]
I agree, let's give it a try!

**Comment by JeanMertz** [comment]
By the way, I was looking for how this would work in practice.

Here's a PR that got merged by bors using the squash strategy (see how it closes the PR instead of merging it, and updates the title):

https://github.com/leanprover-community/mathlib/pull/3448

And here's the final commit that ends up in master:

https://github.com/leanprover-community/mathlib/commit/dced34364276b7036dfe3cbe6b335d97405cc757

And here's their Bors config:

https://github.com/leanprover-community/mathlib/blob/5fd2037ef291b942180b3189d6404355c825b5a7/bors.toml

I also like the `cut_body_after = "---` option to be able to add more content to the PR description that won't end up in the final commit message.

We probably also want to use `use_codeowners = true`.

**Comment by MOZGIII** [comment]
Thanks for sharing this example, I was looking for something like this!
It looks great! I like how they did `bors r-` there to abort the merge mid-flight. :smile:
Also, what's odd is [this page](https://github.com/leanprover-community/mathlib/actions/runs/177957221) shows build took 1:05, but the time between the command to bors to merge, and the success notification comment from bors is ~1:10.
At the same time, there's [another execution](https://github.com/leanprover-community/mathlib/actions/runs/178004238) of the same workflow (that only took 0:31). It is confusing, maybe they have their CI workflow run fully on each push, and also run bors with it's own test runs?

Btw, one of the problems that bors addresses as it's main feature - ["semantic conflicts"](https://bors.tech/essay/2017/02/02/pitch/) is already handled at Github Actions in our setup - the tests are run on the PR integration branch, i.e. a latest `master` with the PR branch merged in. This is why sometimes one might see odd things - like the errors that aren't reproducible locally, but visible at CI - that become visible after merging the latest `master` in locally. Just wanted to point this out, maybe people aren't aware.

**Comment by jamtur01** [comment]
bors try

**Comment by jamtur01** [comment]
bors try

**Comment by jamtur01** [comment]
Okay - I've got this to the point where we can test it. Is everyone okay with running a test for the rest of the sprint? Do y'all want me to talk through the process before we kick off?

**Comment by jamtur01** [comment]
As a placeholder for rollback:

1. Remove bors Github app.
2. Remove Linux unix tests from prs.yml.
3. Leave other jobs unchanged.

**Review by JeanMertz — APPROVED** [review]
This looks sweet 👌

I'm in favor of testing this out on master for a week, and see how we feel about it.

File: .github/workflows/prs.yml
Comment (Hoverbear) [inline .github/workflows/prs.yml:47]: It would be preferable to run these tests via our cross builders so we can test our release process.

File: Makefile
Comment (Hoverbear) [inline Makefile:752]: This might be unintentional

File: scripts/environment/Dockerfile
Comment (Hoverbear) [inline scripts/environment/Dockerfile:13]: I think you can just delete this line no?

**Review by Hoverbear — APPROVED** [review]

**Comment by Hoverbear** [comment]
Let a couple nits but otherwise looks exciting!

File: scripts/environment/Dockerfile
Comment (jamtur01) [reply scripts/environment/Dockerfile:13]: There's a bug in there and the {} don't seem to work right.

File: Makefile
Comment (jamtur01) [reply Makefile:752]: Yeah. Weird.

**Comment by jamtur01** [comment]
bors try

File: .github/workflows/tests.yml
Comment (MOZGIII) [inline .github/workflows/tests.yml:377]: How does this work with bors?

File: .github/workflows/tests.yml
Comment (MOZGIII) [inline .github/workflows/tests.yml:373]: We can combine this task with `build-x86_64-unknown-linux-gnu` to avoid building vector twice.

File: .github/workflows/tests.yml
Comment (jamtur01) [reply .github/workflows/tests.yml:377]: Correct - this wouldn't impact bors.

File: .github/workflows/tests.yml
Comment (MOZGIII) [inline .github/workflows/tests.yml:422]: Will we be able to set the `ci-condition: k8s e2e all targets` with bors? Also, we might not need this anymore, and just run all the target all the time.

File: scripts/environment/Dockerfile
Comment (MOZGIII) [reply scripts/environment/Dockerfile:13]: The `-p` flag on the next line creates the paths recursively as needed, so creating  `/git/timberio/vector/scripts` separately is not required

File: .github/workflows/tests.yml
Comment (jamtur01) [reply .github/workflows/tests.yml:412]: I think we can pull it out. Feel free to remove the logic in `master`.

File: .github/workflows/environment.yml
Comment (MOZGIII) [inline .github/workflows/environment.yml:31]: I wonder why are we adding another tag for the environment image. I was under the impression it was perfectly usable as-is. @Hoverbear wasn't it the intent behind the `environment` image in the first place?

**Review by MOZGIII — COMMENTED** [review]
It looks like this PR is mixing together two concerns: switching to bors, and switching to running off of the environment docker image in CI.
I assume the environment image was included as a trivial addition, but we've tried this before, and I don't recall what the outcome was, but we ended up with doing things directly on the host. Should we, perhaps, split the docker image switch to a different PR, and discuss it separately?

File: .github/CODEOWNERS
Comment (MOZGIII) [inline .github/CODEOWNERS:39]: Nit: we should probably rename this script to `/scripts/environment/setup-helm.sh`, as it's no longer CI-specific, but rather environment-specific.

File: scripts/environment/ci-setup-helm.sh
Comment (MOZGIII) [inline scripts/environment/ci-setup-helm.sh:4]: Note to self: we need to figure out a way to manage this and keep it up to date, how that the files are moving around.

File: Makefile
Comment (MOZGIII) [inline Makefile:989]: Why is this needed? The logic looks rather odd to me, in which cases will it kick in?

Comment (jamtur01) [reply Makefile:989]: We need to cater for two scenarios - sudo locally and no sudo inside our containers.

Comment (MOZGIII) [reply Makefile:989]: Why do we run this inside the container though? We won't be clearing up space due to how docker layers work, and if this is supposed to clear the host paths from inside the container - it won't work either.

File: .github/workflows/tests.yml
Comment (MOZGIII) [reply .github/workflows/tests.yml]: Interesting, why do we then need these `if`s? If I understand correctly, those jobs will only be run by bors now, but if bors doesn't propagate them - what's the point of having such conditions in the first place?

**Comment by jamtur01** [comment]
I've removed all non-bors related changes. They did speed up runs but I can revisit that at a later date.

File: .github/workflows/tests.yml
Comment (jamtur01) [reply .github/workflows/tests.yml]: I didn't remove them because it had no impact on testing bors or not.

File: .github/bors.toml
Comment (MOZGIII) [inline .github/bors.toml:2]: I've seen that bors only posts the link to the `ci` job. I was wondering if it's possible to make it post the link to all the involved runs? I'll make the interesting content a couple clicks closer, and, afaik, it'll also embed the success state of each of the individual job - thus providing us a summary of the test results.

Might this setting affect that?

Comment (jamtur01) [reply .github/bors.toml:2]: No - this setting determines whether the CI passed or failed and whether it can be merged.

**Comment by MOZGIII** [comment]
How about renaming `.github/workflows/tests.yml` to `.github/workflows/bors.yml` for more clarity? Seems like bors doesn't care how we name things from the Github Actions side, and it might simplify the mental model of our test files layout.

File: .github/bors.toml
Comment (MOZGIII) [reply .github/bors.toml:2]: I wonder why it's an array...

Can't we just put the list of all the jobs there:
```toml
"test-linux",
"build-x86_64-unknown-linux-gnu",
"build-aarch64-unknown-linux-gnu",
"test-mac",
"test-windows",
"test-misc",
"test-integration-aws",
"test-integration-clickhouse",
"test-integration-docker",
"test-integration-elasticsearch",
"test-integration-gcp",
"test-integration-humio",
"test-integration-influxdb",
"test-integration-kafka",
"test-integration-loki",
"test-integration-pulsar",
"test-integration-splunk",
"build-deb-package",
"compute-k8s-test-plan",
"test-e2e-kubernetes",
"check-component-features",
```

**Comment by jamtur01** [comment]
I'm going to merge this. From here until we end the test:

Once you’ve reviewed the PR, instead of clicking the green “Merge Button”, leave a comment like this on the pull request:

```text
bors r+
```

Equivalently, you can comment the following:

```text
bors merge
```

The pull request, as well as any other pull requests that are reviewed around the same time, will be merged into a branch called `staging`. CI will run there and report the result back. If that result is “OK”, `master` gets fast-forwarded to reach it.

There’s also:

```text
bors try
```

When this is run, your branch and master get merged into `trying`, and bors will report the results just like the `staging` branch would. Only reviewers can push to this.

**Pull request closed without merging**
